src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -175,7 +180,33 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
                 // for empty array.
                 return false;
             }
-            _ => {
+            ty::Bool
+            | ty::Char
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Str
+            | ty::Never
+            | ty::Error => {
+                // These primitive types are always structural match.
+                //
+                // `Never` is kind of special here, but as it is not inhabitable, this should be fine.
```

Comment:
The key property is, what, that comparing the fields is the same as calling `PartialEq`? That's certainly true for `Never`.

Reply:
As far as I understand this, yes.